### PR TITLE
Adds Equity Fill Model

### PR DESCRIPTION
**Pull request by AlexCatarino**
#### Description
- Adds a fill model tailored for equity;
- Adds new unit tests for the new fill model and fix regression tests;
- Change the default fill model in the Equity class;
- Change the brokerage models to get the fill model according to the security type.

#### Related Issue
Addresses #4567 

#### Motivation and Context
Bug fix

#### How Has This Been Tested?
New unit tests and regression tests,

#### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Common/Orders/Fills/EquityFillModel.cs [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace QuantConnect.Orders.Fills
+{
+    /// <summary>
+    /// Implements <see cref="IFillModel"/> for Equity
+    /// </summary>
+    public class EquityFillModel : FillModel
```
Comment (adam-may) [inline Common/Orders/Fills/EquityFillModel.cs:29]: This is very similar to a Fill Model that we built for another product type, so maybe the term "EquityFillModel" should be changed to describe how the model fills the orders?

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Nice first step 👍 leaving some concerns, requests 😃
*believe there are a few regression algorithms failing with current impl

File: Common/Orders/Fills/EquityFillModel.cs (hunk 1 above)
Comment (Martin-Molinero) [reply Common/Orders/Fills/EquityFillModel.cs:29]: Hey @adam-may! Curious which product type? The current plan as seen in the `DefaultBrokerageModel` is to create a fill model per security type but they could potentially inherit from each other is it's correct

File: Common/Orders/Fills/EquityFillModel.cs [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+                if (quoteBar != null)
+                {
+                    return quoteBar;
+                }
+            }
+
+            if (subscribedTypes.Contains(typeof(Tick)))
+            {
+                var quote = asset.Cache.GetAll<Tick>().LastOrDefault(x => x.TickType == TickType.Quote);
+                if (quote != null)
```
[the remark this answers is not in the document]
Comment (Martin-Molinero) [reply Common/Orders/Fills/EquityFillModel.cs:632]: hm I thought I was wrong above but for example if `QuoteBar` resolution is daily and we have tick resolution subscription most of the day we would be using old data? same for `TradeBar` 🤔, note that `master` is using `Ticks` first too

File: Common/Orders/Fills/EquityFillModel.cs [hunk 3]
```diff
@@ -125,10 +126,18 @@ public override OrderEvent StopMarketFill(Security asset, StopMarketOrder order)
                     // Buy Stop: If Price Above set point, Buy:
                     if (lastTradeBar.High >= order.StopPrice)
                     {
-                        DateTime unused;
+                        // Bar opens above stop price, fill at open price
+                        if (lastTradeBar.Open >= order.StopPrice)
+                        {
+                            fill.FillPrice = lastTradeBar.Open + slip;
```
Comment (Martin-Molinero) [inline Common/Orders/Fills/EquityFillModel.cs:132]: Sweet 👍

**Review by Martin-Molinero — APPROVED** [review]

**Review by Martin-Molinero — COMMENTED** [review]
`StopLossOnOrderEventRegressionAlgorithm` & `UpdateOrderRegressionAlgorithm` seem to be failing @AlexCatarino

**Pull request closed without merging**